core/arch/arm/kernel/thread_spmc_a64.S

```diff
@@ -171,13 +185,32 @@ END_FUNC thread_rpc
  * The current thread as indicated by @thread_index has just been
  * suspended.  The job here is just to inform normal world the thread id to
  * resume when returning.
+ * If the active FF-A endpoint is OP-TEE (or a TA) then an this function send an
+ * OPTEE_FFA_YIELDING_CALL_RETURN_INTERRUPT message to the normal world via the
+ * FFA_MSG_SEND_DIRECT_RESP interface. This is handled by the OP-TEE
+ * driver in Linux so it can schedule task to the thread.
+ * If the active endpoint is an SP the function sends an FFA_INTERRUPT. This is
+ * handled by the FF-A driver and after taking care of the NWd interrupts it
+ * returns via an FFA_RUN call.
+ * The active endpoint is determined by checking the THREAD_FLAGS_FFA_ONLY flag
+ * in threads[w0].flags. This is only set for the thread which handles SPs.
  */
 FUNC thread_foreign_intr_exit , :
-	/* load threads[w0].tsd.rpc_target_info into w1 */
 	mov	x1, #THREAD_CTX_SIZE
 	adr_l	x2, threads
 	madd	x1, x1, x0, x2
+#if CFG_SECURE_PARTITION
+	/* load threads[w0].flags into w2 */
+	ldr	w2, [x1, #THREAD_CTX_FLAGS]
+	/* load threads[w0].tsd.rpc_target_info into w1 */
+	ldr	w1, [x1, #THREAD_CTX_TSD_RPC_TARGET_INFO]
+	and     w2, w2, #THREAD_FLAGS_FFA_ONLY
+	cmp	x2, 0
+	bne	thread_ffa_interrupt
+#else
+	/* load threads[w0].tsd.rpc_target_info into w1 */
 	ldr	w1, [x1, #THREAD_CTX_TSD_RPC_TARGET_INFO]
```

Comment:
I'd rather not do this in an `#else`, please refactor to be able to initialize `w1` before the `#if` above.

Reply:
I've refactored this part.